File: android/app/src/main/java/com/todo/app/data/model/StatsTimeline.kt
package com.todo.app.data.model

import java.time.*

data class SubtaskEvent(val todo: Todo, val subtask: Subtask, val date: LocalDate, val time: Instant?)
data class TimerArc(val part: LearningPart, val startMinute: Float, val endMinute: Float)
data class TimelineCompletion(val todo: Todo, val date: LocalDate, val time: Instant?)

object StatsTimeline {
    fun days(period: String, target: LocalDate): List<LocalDate> {
        val (start, end) = Learning.range(period, target)
        return generateSequence(start) { it.plusDays(1) }.takeWhile { it < end }.toList()
    }

    fun completions(todos: List<Todo>, period: String, target: LocalDate, zone: ZoneId = ZoneId.systemDefault()): List<TimelineCompletion> {
        val (start, end) = Learning.range(period, target)
        return todos.filterNot { it.deleted }.flatMap { todo ->
            val times = if (todo.taskType == TaskType.WEEKLY_CHECKIN || todo.taskType == TaskType.MONTHLY_CHECKIN) todo.completedDates.distinct()
                else if (todo.completed) listOfNotNull(todo.completedAt) else emptyList()
            times.mapNotNull { value ->
                val time = Learning.instant(value)
                val date = time?.atZone(zone)?.toLocalDate() ?: runCatching { LocalDate.parse(value) }.getOrNull() ?: return@mapNotNull null
                if (date >= start && date < end) TimelineCompletion(todo, date, time) else null
            }
        }
    }

    fun hitSegments(parts: List<TimerArc>, date: LocalDate, minute: Float, tolerance: Float = 0f): List<TimerArc> =
        parts.filter { it.part.date == date && minute >= it.startMinute - tolerance && minute <= it.endMinute + tolerance }

    // 任务与子步骤共用精确本地时间刻度，保留秒及小数秒。
    fun clockMinute(time: Instant, zone: ZoneId = ZoneId.systemDefault()): Double =
        time.atZone(zone).toLocalTime().toNanoOfDay() / 60_000_000_000.0

    fun subtasks(todos: List<Todo>, period: String, target: LocalDate, zone: ZoneId = ZoneId.systemDefault()): List<SubtaskEvent> {
        val (start, end) = Learning.range(period, target)
        return todos.filterNot { it.deleted }.flatMap { todo -> todo.subtasks.mapNotNull { s ->
            if (!s.completed || s.completedAt.isNullOrBlank()) return@mapNotNull null
            val time = Learning.instant(s.completedAt)
            val date = time?.atZone(zone)?.toLocalDate() ?: runCatching { LocalDate.parse(s.completedAt) }.getOrNull() ?: return@mapNotNull null
            if (date >= start && date < end) SubtaskEvent(todo, s, date, time) else null
        } }
    }
    fun arcs(entries: List<TimeEntry>, source: TaskReference, period: String, target: LocalDate, zone: ZoneId = ZoneId.systemDefault()): List<TimerArc> {
        val (start, end) = Learning.range(period, target)
        val conflicts = Learning.overlaps(entries)
        return entries.filter { it.id !in conflicts && it.task_ref.source_type == source.source_type && it.task_ref.source_id == source.source_id }
            .flatMap { Learning.split(it, zone) }.filter { it.date >= start && it.date < end }.flatMap { part ->
                val result = mutableListOf<TimerArc>(); var cursor = part.startedAt
                while (cursor < part.endedAt) {
                    val transition = zone.rules.nextTransition(cursor)?.instant
                    val next = if (transition != null && transition < part.endedAt) transition else part.endedAt
                    val minute = clockMinute(cursor, zone)
                    result.add(TimerArc(part, minute.toFloat(), minOf(1440.0, minute + Duration.between(cursor, next).toMillis() / 60000.0).toFloat()))
                    cursor = next
                }
                result
            }
    }
    fun hitArcs(arcs: List<TimerArc>, minute: Float, tolerance: Float = 0f): List<TimerArc> = arcs.filter { a ->
        listOf(minute, minute + 1440, minute - 1440).any { it >= a.startMinute - tolerance && it <= a.endMinute + tolerance }
    }
}

File: windows/src/statsTimeline.js
import { localDay, learningRange, splitTimeEntry, overlappingEntries } from './timeTracking.js';

// 所有钟面标记共用本地时间刻度，保留秒和毫秒，不添加角度偏移。
export function clockMinute(value) {
    const d = new Date(value);
    return d.getHours() * 60 + d.getMinutes() + d.getSeconds() / 60 + d.getMilliseconds() / 60000;
}

// 悬浮提示用紧凑的本地日期与分钟；原始记录及钟面精度保持不变。
export function clockTooltipDateTime(value) {
    if (typeof value === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(value)) return { date: value.slice(5), time: '' };
    const d = new Date(value), pad = n => String(n).padStart(2, '0');
    return { date: `${pad(d.getMonth() + 1)}-${pad(d.getDate())}`, time: `${pad(d.getHours())}:${pad(d.getMinutes())}` };
}

export function collectSubtaskEvents(todos, period, target) {
    const { start, end } = learningRange(period, target);
    return todos.filter(t => !t.deleted).flatMap(todo => (todo.subtasks || []).flatMap(subtask => {
        if (!subtask.completed || !subtask.completed_at) return [];
        const time = subtask.completed_at;
        if (!Number.isFinite(Date.parse(time))) return [];
        const explicit = time.includes('T');
        if (!explicit && (!/^\d{4}-\d{2}-\d{2}$/.test(time) || new Date(time).toISOString().slice(0, 10) !== time)) return [];
        const date = explicit ? localDay(time) : time;
        return date >= start && date < end ? [{ id: `${todo.id}:${subtask.id}`, todo, subtask, date, completed_at: time, explicit }] : [];
    }));
}

// 在偏移变化处分段：钟面角度用当地时间，时长仍用真实时间差。
export function clockSegments(part) {
    const result = []; let cursor = part.started_at;
    while (cursor < part.ended_at) {
        const offset = new Date(cursor).getTimezoneOffset();
        let next = Math.min(cursor + 60000, part.ended_at);
        while (next < part.ended_at && new Date(next).getTimezoneOffset() === offset) next = Math.min(next + 60000, part.ended_at);
        if (new Date(next).getTimezoneOffset() !== offset) {
            let low = Math.max(cursor, next - 60000), high = next;
            while (high - low > 1) { const mid = Math.floor((low + high) / 2); if (new Date(mid).getTimezoneOffset() === offset) low = mid; else high = mid; }
            next = high;
        }
        const startMinute = clockMinute(cursor);
        result.push({ ...part, startMinute, endMinute: Math.min(1440, startMinute + (next - cursor) / 60000) });
        cursor = next;
    }
    return result;
}

export function collectTimerArcs(entries, source, period, target) {
    const { start, end } = learningRange(period, target), conflicts = overlappingEntries(entries);
    return entries.filter(e => !conflicts.has(e.id) && e.task_ref.source_type === source.type &&
        (e.task_ref.source_id ?? null) === (source.type === 'personal' ? null : source.id))
        .flatMap(splitTimeEntry).filter(p => p.date >= start && p.date < end).flatMap(clockSegments);
}

export function arcPath(cx, cy, radius, startMinute, endMinute) {
    const point = m => [cx + radius * Math.sin(m / 1440 * 2 * Math.PI), cy - radius * Math.cos(m / 1440 * 2 * Math.PI)];
    const a = point(startMinute), b = point(endMinute);
    if (endMinute - startMinute >= 1440) { const mid = point(startMinute + 720); return `M ${a} A ${radius},${radius} 0 0 1 ${mid} A ${radius},${radius} 0 0 1 ${a}`; }
    return `M ${a} A ${radius},${radius} 0 ${endMinute - startMinute > 720 ? 1 : 0} 1 ${b}`;
}

export function hitArcs(arcs, minute, tolerance = 0) {
    return arcs.filter(a => [minute, minute + 1440, minute - 1440].some(m => m >= a.startMinute - tolerance && m <= a.endMinute + tolerance));
}
